x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/OptimizerRules.java

```diff
@@ -220,178 +180,6 @@ protected Expression maybeSimplifyNegatable(Expression e) {
-
-        @Override
-        protected LogicalPlan rule(Filter filter) {
-            Expression condition = filter.condition().transformUp(BinaryLogic.class, PruneFilters::foldBinaryLogic);
-
-            if (condition instanceof Literal) {
-                if (TRUE.equals(condition)) {
-                    return filter.child();
-                }
-                if (FALSE.equals(condition) || Expressions.isNull(condition)) {
-                    return skipPlan(filter);
-                }
-            }
-
-            if (condition.equals(filter.condition()) == false) {
-                return new Filter(filter.source(), filter.child(), condition);
-            }
-            return filter;
-        }
-
-        protected abstract LogicalPlan skipPlan(Filter filter);
-
-        private static Expression foldBinaryLogic(BinaryLogic binaryLogic) {
-            if (binaryLogic instanceof Or or) {
-                boolean nullLeft = Expressions.isNull(or.left());
-                boolean nullRight = Expressions.isNull(or.right());
-                if (nullLeft && nullRight) {
-                    return new Literal(binaryLogic.source(), null, DataType.NULL);
-                }
-                if (nullLeft) {
-                    return or.right();
-                }
-                if (nullRight) {
-                    return or.left();
-                }
-            }
-            if (binaryLogic instanceof And and) {
-                if (Expressions.isNull(and.left()) || Expressions.isNull(and.right())) {
-                    return new Literal(binaryLogic.source(), null, DataType.NULL);
-                }
-            }
-            return binaryLogic;
-        }
-    }
-
-    // NB: it is important to start replacing casts from the bottom to properly replace aliases
-    public abstract static class PruneCast<C extends Expression> extends Rule<LogicalPlan, LogicalPlan> {
-
-        private final Class<C> castType;
-
-        public PruneCast(Class<C> castType) {
-            this.castType = castType;
-        }
-
-        @Override
-        public final LogicalPlan apply(LogicalPlan plan) {
-            return rule(plan);
-        }
-
-        protected final LogicalPlan rule(LogicalPlan plan) {
-            // eliminate redundant casts
-            return plan.transformExpressionsUp(castType, this::maybePruneCast);
-        }
-
-        protected abstract Expression maybePruneCast(C cast);
-    }
-
-    public abstract static class SkipQueryOnLimitZero extends OptimizerRule<Limit> {
-        @Override
-        protected LogicalPlan rule(Limit limit) {
-            if (limit.limit().foldable()) {
-                if (Integer.valueOf(0).equals((limit.limit().fold()))) {
-                    return skipPlan(limit);
-                }
-            }
-            return limit;
-        }
-
-        protected abstract LogicalPlan skipPlan(Limit limit);
-    }
-
     public static class FoldNull extends OptimizerExpressionRule<Expression> {
```

Comment:
To be consistent with the other optimization rules, maybe we should also extract the rules in this class into their own files and place them into `x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/`.

Reply:
I agree, but I'm going to save that work for another time so I can get this in.